File: objectiveai-sdk-js/src/functions/inventions/recursive/http.ts
import z from "zod";
import { ObjectiveAI, type RequestOptions } from "../../../client";
import { Stream } from "../../../stream";
import {
  FunctionsInventionsRecursiveRequestFunctionInventionRecursiveCreateParamsSchema,
  type FunctionsInventionsRecursiveRequestFunctionInventionRecursiveCreateParams,
} from "./request/functionInventionRecursiveCreateParams";
import type { FunctionsInventionsRecursiveResponseUnaryFunctionInventionRecursive } from "./response/unary/functionInventionRecursive";
import type { FunctionsInventionsRecursiveResponseStreamingFunctionInventionRecursiveChunk } from "./response/streaming/functionInventionRecursiveChunk";

export const FunctionsInventionsRecursiveRequestFunctionInventionRecursiveCreateParamsStreamingSchema =
  FunctionsInventionsRecursiveRequestFunctionInventionRecursiveCreateParamsSchema.extend({
    stream: z.literal(true),
  });
export type FunctionsInventionsRecursiveRequestFunctionInventionRecursiveCreateParamsStreaming = z.infer<
  typeof FunctionsInventionsRecursiveRequestFunctionInventionRecursiveCreateParamsStreamingSchema
>;

export const FunctionsInventionsRecursiveRequestFunctionInventionRecursiveCreateParamsUnarySchema =
  FunctionsInventionsRecursiveRequestFunctionInventionRecursiveCreateParamsSchema.extend({
    stream: z.literal(false).optional().nullable(),
  });
export type FunctionsInventionsRecursiveRequestFunctionInventionRecursiveCreateParamsUnary = z.infer<
  typeof FunctionsInventionsRecursiveRequestFunctionInventionRecursiveCreateParamsUnarySchema
>;

export function functionsInventionsRecursiveCreateFunctionInventionRecursive(
  client: ObjectiveAI,
  body: FunctionsInventionsRecursiveRequestFunctionInventionRecursiveCreateParamsStreaming,
  options?: RequestOptions,
): Promise<Stream<FunctionsInventionsRecursiveResponseStreamingFunctionInventionRecursiveChunk>>;
export function functionsInventionsRecursiveCreateFunctionInventionRecursive(
  client: ObjectiveAI,
  body: FunctionsInventionsRecursiveRequestFunctionInventionRecursiveCreateParamsUnary,
  options?: RequestOptions,
): Promise<FunctionsInventionsRecursiveResponseUnaryFunctionInventionRecursive>;
export function functionsInventionsRecursiveCreateFunctionInventionRecursive(
  client: ObjectiveAI,
  body: FunctionsInventionsRecursiveRequestFunctionInventionRecursiveCreateParams,
  options?: RequestOptions,
): Promise<
  | Stream<FunctionsInventionsRecursiveResponseStreamingFunctionInventionRecursiveChunk>
  | FunctionsInventionsRecursiveResponseUnaryFunctionInventionRecursive
> {
  if (body.stream) {
    return client.post_streaming<FunctionsInventionsRecursiveResponseStreamingFunctionInventionRecursiveChunk>(
      "/functions/inventions/recursive",
      body,
      options,
    );
  }
  return client.post_unary<FunctionsInventionsRecursiveResponseUnaryFunctionInventionRecursive>(
    "/functions/inventions/recursive",
    body,
    options,
  );
}

File: objectiveai-sdk-py/objectiveai_sdk/functions/inventions/recursive/http.py
"""HTTP functions for recursive function inventions."""

from __future__ import annotations

from typing import TYPE_CHECKING, Union

if TYPE_CHECKING:
    from objectiveai.client import ObjectiveAI
    from objectiveai.functions.inventions.recursive.request import (
        FunctionInventionRecursiveCreateParams,
    )
    from objectiveai.functions.inventions.recursive.response.streaming import (
        FunctionInventionRecursiveChunk,
    )
    from objectiveai.functions.inventions.recursive.response.unary import (
        FunctionInventionRecursive,
    )
    from objectiveai.stream import Stream


async def create_function_invention_recursive(
    client: ObjectiveAI,
    params: FunctionInventionRecursiveCreateParams,
) -> Union[FunctionInventionRecursive, Stream[FunctionInventionRecursiveChunk]]:
    """Create a recursive function invention.

    If ``params.stream`` is true, returns a streaming response.
    """
    if getattr(params, "stream", None):
        return await client.post_streaming(
            "functions/inventions/recursive", params,
        )
    return await client.post_unary("functions/inventions/recursive", params)
